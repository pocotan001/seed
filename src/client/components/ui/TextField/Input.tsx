import styled from "styled-components";
import { Color } from "../../styles/enums";

const Input = styled("input")`
  font-size: 1rem;
  line-height: 1.5;
  display: block;
  padding: 0.4rem 0.75rem;
  width: 100%;
  max-width: 100%;
  color: ${Color.Grey800};
  border: 1px solid ${Color.Grey400};
  border-radius: 3px;
  background: ${Color.White};

  &:hover {
    border-color: ${Color.Grey500};
  }

  &:disabled {
    cursor: default;
    opacity: 0.6;
    border-color: ${Color.Grey400};
  }

  &[aria-invalid="true"] {
    border-color: ${Color.Pink500};
  }

  &::-webkit-input-placeholder {
    color: ${Color.Grey400};
  }

  &::-moz-placeholder {
    color: ${Color.Grey400};
  }
`;

export default Input;
