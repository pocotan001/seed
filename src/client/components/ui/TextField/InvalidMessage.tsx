import styled from "styled-components";
import { Color } from "../../styles/enums";

const InvalidMessage = styled("p").attrs({
  role: "alert"
})`
  font-size: 80%;
  color: ${Color.Pink500};
  white-space: pre;
  margin-top: 4px;
`;

export default InvalidMessage;
