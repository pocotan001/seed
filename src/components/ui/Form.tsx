import { FORM_ERROR, FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

export type IFormOnSubmit = (
  values: IFormValues,
  form: FormApi
) => Promise<void | string> | void | string;

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: IFormOnSubmit;
      }
    > {}

const focusOnErrors = createDecorator();

export default class Form extends React.PureComponent<IFormProps> {
  handleSubmit = async (values: IFormValues, form: FormApi): Promise<any> => {
    const { onSubmit } = this.props;
    const err = await onSubmit(values, form);

    if (err) {
      return { [FORM_ERROR]: err };
    }
  };

  render() {
    const { onSubmit, ...rest } = this.props;

    return (
      <FinalForm
        decorators={[focusOnErrors]}
        onSubmit={this.handleSubmit}
        {...rest}
      />
    );
  }
}
