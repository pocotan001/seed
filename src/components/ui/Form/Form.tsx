import { FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

interface ISubmitErrors {
  [name: string]: string;
}

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: (
          values: IFormValues,
          form: FormApi,
          callback?: (errors?: ISubmitErrors) => void
        ) => Promise<ISubmitErrors | void> | ISubmitErrors | void;
      }
    > {}

const focusOnErrors = createDecorator();
const Form: React.SFC<IFormProps> = props => (
  <FinalForm decorators={[focusOnErrors]} {...props} />
);

export default Form;
