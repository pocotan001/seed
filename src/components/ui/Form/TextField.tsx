import * as React from "react";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import { ITheme } from "~/components/styles/theme";
import styled, { css } from "~/components/styles/themedStyledComponents";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";

type IInputAttributes = React.InputHTMLAttributes<HTMLInputElement>;
export type ITextFieldStyleProps = IMarginProps;

interface IAdditionalProps extends ITextFieldStyleProps {
  theme?: ITheme;
}

const TextField: React.SFC<
  IFieldRenderProps<IInputAttributes> & IAdditionalProps
> = ({ input, meta, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return <input {...input} {...rest} aria-invalid={isInvalid} />;
};

export const textFieldStyles = css<ITextFieldStyleProps>`
  font-size: 1rem;
  line-height: 1.5;
  display: block;
  padding: 0.4em 0.75em;
  width: 100%;
  max-width: 100%;
  color: ${({ theme }) => theme.colors.grey800};
  border: 1px solid ${({ theme }) => theme.colors.grey400};
  border-radius: 3px;
  background: ${({ theme }) => theme.colors.white};

  &:hover {
    border-color: ${({ theme }) => theme.colors.grey500};
  }

  &:disabled {
    cursor: default;
    opacity: 0.6;
    border-color: ${({ theme }) => theme.colors.grey400};
  }

  &[aria-invalid="true"] {
    border-color: ${({ theme }) => theme.colors.pink500};
  }

  &::-webkit-input-placeholder {
    color: ${({ theme }) => theme.colors.grey400};
  }

  &::-moz-placeholder {
    color: ${({ theme }) => theme.colors.grey400};
  }

  ${margin};
`;

const StyledTextField = styled(TextField)`
  ${textFieldStyles};
`;

const AdaptedTextField: React.SFC<
  IFieldProps<IInputAttributes> & IAdditionalProps
> = props => <Field {...props} component={StyledTextField} />;

AdaptedTextField.defaultProps = {
  type: "text"
};

export default AdaptedTextField;
