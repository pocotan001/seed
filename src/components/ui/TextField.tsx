import * as React from "react";
import styled from "styled-components";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import { colors } from "~/components/styles/theme";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";

interface IInputAttributes extends React.InputHTMLAttributes<HTMLInputElement> {
  // Text field should have autocomplete attribute
  // https://www.chromium.org/developers/design-documents/create-amazing-password-forms#TOC-Use-autocomplete-attributes
  // https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete#Values
  autoComplete: string;
}

export type ITextFieldStyleProps = IMarginProps;

interface ITextFieldProps
  extends IFieldRenderProps<IInputAttributes>,
    ITextFieldStyleProps {}

export const TextField = styled<ITextFieldProps>(
  ({ input, meta, m, mt, mr, mb, ml, ...rest }) => (
    <input
      {...input}
      {...rest}
      aria-invalid={Boolean(meta.touched && meta.error)}
    />
  )
)`
  font-size: 1rem;
  line-height: 1.5;
  display: block;
  padding: 0.4em 0.75em;
  width: 100%;
  max-width: 100%;
  color: ${colors.grey800};
  border: 1px solid ${colors.grey400};
  border-radius: 3px;
  background: ${colors.white};

  &:hover {
    border-color: ${colors.grey500};
  }

  &:disabled {
    cursor: default;
    opacity: 0.6;
    border-color: ${colors.grey400};
  }

  &[aria-invalid="true"] {
    border-color: ${colors.pink500};
  }

  &::-webkit-input-placeholder {
    color: ${colors.grey400};
  }

  &::-moz-placeholder {
    color: ${colors.grey400};
  }

  ${margin};
`;

const AdaptedTextField: React.SFC<
  IFieldProps<IInputAttributes> & ITextFieldStyleProps
> = props => <Field {...props} component={TextField as any} />;

AdaptedTextField.defaultProps = {
  type: "text"
};

export default AdaptedTextField;
