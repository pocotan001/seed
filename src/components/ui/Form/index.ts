import AutoSave from "./AutoSave";
import CheckBox from "./CheckBox";
import _Form from "./Form";
import Radio from "./Radio";
import Select from "./Select";
import SubmitError from "./SubmitError";
import TextArea from "./TextArea";
import TextField from "./TextField";
import ValidationError from "./ValidationError";

const Form: typeof _Form & {
  AutoSave: typeof AutoSave;
  CheckBox: typeof CheckBox;
  Radio: typeof Radio;
  Select: typeof Select;
  SubmitError: typeof SubmitError;
  TextArea: typeof TextArea;
  TextField: typeof TextField;
  ValidationError: typeof ValidationError;
} = _Form as any;

Form.AutoSave = AutoSave;
Form.CheckBox = CheckBox;
Form.Radio = Radio;
Form.Select = Select;
Form.SubmitError = SubmitError;
Form.TextArea = TextArea;
Form.TextField = TextField;
Form.ValidationError = ValidationError;

export default Form;
