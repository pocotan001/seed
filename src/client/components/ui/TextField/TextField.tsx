import React from "react";
import { WrappedFieldProps } from "redux-form";
import styled from "styled-components";
import margin, { MarginProps } from "../../styles/extends/margin";
import Input from "./Input";
import InvalidMessage from "./InvalidMessage";
import Label from "./Label";

interface OwnProps {
  label?: string;
  className?: string;
}

type TextFieldProps = OwnProps &
  React.InputHTMLAttributes<HTMLInputElement> &
  WrappedFieldProps &
  MarginProps;

const TextField = ({
  label,
  className,
  input,
  meta: { touched, error },
  ...rest
}: TextFieldProps) => {
  const isInvalid = touched && !!error;

  return (
    <div className={className}>
      {label && (
        <Label htmlFor={rest.id}>
          {label}
          {rest.required && " *"}
        </Label>
      )}
      <Input aria-invalid={isInvalid} {...input} {...rest} />
      {isInvalid && <InvalidMessage>{error}</InvalidMessage>}
    </div>
  );
};

export default styled(TextField)`
  font-size: 1rem;

  ${margin};
` as typeof TextField;
