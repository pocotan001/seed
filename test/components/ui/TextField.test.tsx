import * as React from "react";
import * as renderer from "react-test-renderer";
import Form from "~/components/ui/Form";
import TextField from "~/components/ui/TextField";

const handleSubmit = () => undefined;

describe("<TextField>", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <TextField name="test" autoComplete="off" />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
