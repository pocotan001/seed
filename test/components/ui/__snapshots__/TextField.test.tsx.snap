// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField> renders correctly 1`] = `
<input
  aria-invalid={false}
  className="sc-bdVaJa chZRqM"
  name="textfield"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value=""
/>
`;
